File: src/main/java/io/github/chindeaone/collectiontracker/api/bazaarapi/FetchBazaarPrice.java
package io.github.chindeaone.collectiontracker.api.bazaarapi;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.github.chindeaone.collectiontracker.api.URLManager;
import io.github.chindeaone.collectiontracker.api.tokenapi.TokenManager;
import io.github.chindeaone.collectiontracker.collections.BazaarCollectionsManager;
import io.github.chindeaone.collectiontracker.collections.CollectionsManager;
import io.github.chindeaone.collectiontracker.collections.prices.GemstonePrices;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.*;
import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;

import static io.github.chindeaone.collectiontracker.api.URLManager.HTTP_CLIENT;

public class FetchBazaarPrice {

    private static final Logger logger = LogManager.getLogger(FetchBazaarPrice.class);

    public static void fetchData(String uuid, String token, String collection) {
        try {
            URI uri = URI.create(URLManager.BAZAAR_URL);

            HttpRequest request = HttpRequest.newBuilder(uri)
                    .timeout(Duration.ofSeconds(5))
                    .header("X-UUID", uuid)
                    .header("Authorization", "Bearer " + token)
                    .header("X-COLLECTION", collection)
                    .header("User-Agent", URLManager.AGENT)
                    .header("Accept", "application/json")
                    .GET()
                    .build();

            HttpResponse<InputStream> response =
                    HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofInputStream());

            int status = response.statusCode();

            if (status == 401) {
                logger.warn("[SCT]: Invalid or expired token. Fetching a new one and retrying...");
                TokenManager.fetchAndStoreToken();
                token = TokenManager.getToken(); // get the new token

                request = HttpRequest.newBuilder(uri)
                        .timeout(Duration.ofSeconds(5))
                        .header("X-UUID", uuid)
                        .header("Authorization", "Bearer " + token)
                        .header("X-COLLECTION", collection)
                        .header("User-Agent", URLManager.AGENT)
                        .header("Accept", "application/json")
                        .GET()
                        .build();
                response = HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofInputStream());
                status = response.statusCode();
            }

            if (status == 200) {
                try (Reader reader = new InputStreamReader(response.body(), StandardCharsets.UTF_8)) {
                    JsonObject jsonObject = JsonParser.parseReader(reader).getAsJsonObject();

                    // Map the new data structure
                    Map.Entry<String, JsonElement> entry = jsonObject.entrySet().iterator().next();

                    String data = entry.getValue().getAsJsonObject().toString();
                    String type = entry.getKey();

                    if ("gemstone".equals(type)) {
                        GemstonePrices.setPrices(data);
                    } else {
                        BazaarCollectionsManager.setPricesAndRecipes(data, type);
                    }
                    CollectionsManager.collectionType = type;
                    logger.info("[SCT]: Successfully fetched bazaar price for collection '{}'", collection);
                }
            }
        } catch (Exception e) {
            logger.error("[SCT]: Error fetching bazaar price for collection '{}': {}", collection, e.getMessage());
        }
    }

    public static void fetchData(String uuid, String token, List<String> collections) {
        try {
            URI uri = URI.create(URLManager.BAZAAR_URL);

            HttpRequest request = HttpRequest.newBuilder(uri)
                    .timeout(Duration.ofSeconds(5))
                    .header("X-UUID", uuid)
                    .header("Authorization", "Bearer " + token)
                    .header("X-COLLECTION", String.join(", ", collections))
                    .header("User-Agent", URLManager.AGENT)
                    .header("Accept", "application/json")
                    .GET()
                    .build();

            HttpResponse<InputStream> response =
                    HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofInputStream());

            int status = response.statusCode();

            if (status == 401) {
                logger.warn("[SCT]: Invalid or expired token. Fetching a new one and retrying...");
                TokenManager.fetchAndStoreToken();
                token = TokenManager.getToken(); // get the new token

                request = HttpRequest.newBuilder(uri)
                        .timeout(Duration.ofSeconds(5))
                        .header("X-UUID", uuid)
                        .header("Authorization", "Bearer " + token)
                        .header("X-COLLECTION", String.join(", ", collections))
                        .header("User-Agent", URLManager.AGENT)
                        .header("Accept", "application/json")
                        .GET()
                        .build();
                response = HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofInputStream());
                status = response.statusCode();
            }

            if (status == 200) {
                try (Reader reader = new InputStreamReader(response.body(), StandardCharsets.UTF_8)) {
                    JsonObject jsonObject = JsonParser.parseReader(reader).getAsJsonObject();

                    for (Map.Entry<String, JsonElement> collectionEntry : jsonObject.entrySet()) {
                        String collectionId = collectionEntry.getKey();
                        JsonObject typeWrapper = collectionEntry.getValue().getAsJsonObject();

                        if (typeWrapper.entrySet().iterator().hasNext()) {
                            Map.Entry<String, JsonElement> typeEntry = typeWrapper.entrySet().iterator().next();
                            String type = typeEntry.getKey();
                            String data = typeEntry.getValue().getAsJsonObject().toString();

                            if ("gemstone".equals(type)) {
                                GemstonePrices.setPrices(collectionId, data);
                            } else {
                                BazaarCollectionsManager.setPricesAndRecipes(collectionId, data, type);
                            }
                            CollectionsManager.multiCollectionTypes.put(collectionId, type);
                        }
                    }
                    logger.info("[SCT]: Successfully fetched bazaar price for collection list '{}'", collections);
                }
            }
        } catch (Exception e) {
            logger.error("[SCT]: Error fetching bazaar price for collections '{}': {}", collections, e.getMessage());
        }
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/api/bazaarapi/FetchBazaarPrice.kt
package io.github.chindeaone.collectiontracker.api.bazaarapi

import com.google.gson.JsonParser
import io.github.chindeaone.collectiontracker.api.ApiManager
import io.github.chindeaone.collectiontracker.api.tokenapi.TokenManager
import io.github.chindeaone.collectiontracker.collections.BazaarCollectionsManager
import io.github.chindeaone.collectiontracker.collections.CollectionsManager
import io.github.chindeaone.collectiontracker.collections.prices.GemstonePrices
import io.github.chindeaone.collectiontracker.utils.PlayerData
import org.apache.logging.log4j.LogManager
import java.util.concurrent.CompletableFuture

object FetchBazaarPrice {

    private val logger = LogManager.getLogger(FetchBazaarPrice::class.java)

    fun fetchData(collection: String): CompletableFuture<Void> {
        return requestHelper(collection)
            .thenAccept { response ->
                if (response.statusCode() != 200) {
                    logger.error("[SCT]: Failed to fetch bazaar price for collection '{}', response code: {}", collection, response.statusCode())
                    return@thenAccept
                }

                val jsonObject = JsonParser.parseString(response.body()).asJsonObject

                val collectionObject = jsonObject.entrySet().first().value.asJsonObject
                val entry = collectionObject.entrySet().first()

                val type = entry.key
                val data = entry.value.asJsonObject.toString()

                if (type == "gemstone") {
                    GemstonePrices.setPrices(data)
                } else {
                    BazaarCollectionsManager.setPricesAndRecipes(data, type)
                }

                CollectionsManager.collectionType = type
                logger.info("[SCT]: Successfully fetched bazaar price for collection '{}'", collection)
            }
            .exceptionally { e ->
                logger.error("[SCT]: Error fetching bazaar price for collection '{}': {}", collection, e.message)
                null
            }
    }

    fun fetchData(collections: List<String>): CompletableFuture<Void> {
        return requestHelper(addGemstones(collections).joinToString(", "))
            .thenAccept { response ->
                if (response.statusCode() != 200) {
                    logger.error("[SCT]: Failed to fetch bazaar price for collections '{}', response code: {}", collections, response.statusCode())
                    return@thenAccept
                }

                val jsonObject = JsonParser.parseString(response.body()).asJsonObject

                for ((collectionId, wrapperElement) in jsonObject.entrySet()) {
                    val typeWrapper = wrapperElement.asJsonObject

                    val typeEntry = typeWrapper.entrySet().firstOrNull() ?: continue

                    val type = typeEntry.key
                    val data = typeEntry.value.asJsonObject.toString()

                    if (type == "gemstone") {
                        GemstonePrices.setPrices(collectionId, data)
                    } else {
                        BazaarCollectionsManager.setPricesAndRecipes(collectionId, data, type)
                    }

                    CollectionsManager.multiCollectionTypes[collectionId] = type
                }

                logger.info("[SCT]: Successfully fetched bazaar price for collection list '{}'", collections)
            }.exceptionally { e ->
                logger.error("[SCT]: Error fetching bazaar price for collections '{}': {}", collections, e.message)
                null
            }
    }

    private fun requestHelper(collection: String) =
        ApiManager.requestAsync(
            "bazaar",
            mapOf(
                "Authorization" to "Bearer ${TokenManager.token}",
                "X-UUID" to PlayerData.playerUUID,
                "X-COLLECTION" to collection
            )
        )

    private fun addGemstones(collections: List<String>): List<String> {
        val result = collections.toMutableList()

        if ("gemstone" in result) {
            result.remove("gemstone")
            result += listOf(
                "ruby",
                "sapphire",
                "topaz",
                "amethyst",
                "jade",
                "jasper",
                "amber",
                "opal",
                "aquamarine",
                "peridot",
                "citrine",
                "onyx"
            )
        }

        return result
    }
}